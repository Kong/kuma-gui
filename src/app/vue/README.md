---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

