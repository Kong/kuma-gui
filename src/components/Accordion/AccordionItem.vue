<template>
  <li :class="accordionItemClasses">
    <div
      class="accordion-item-header"
      @click="open"
    >
      <!-- This slot will display header -->
      <slot name="accordion-header" />
    </div>

    <transition
      name="accordion"
      @enter="start"
      @after-enter="end"
      @before-leave="start"
    >
      <div v-if="visible">
        <!-- This slot will display whole content -->
        <slot name="accordion-content" />
      </div>
    </transition>
  </li>
</template>

<script>
export default {
  name: 'AccordionItem',
  inject: ['parentAccordion'],
  data() {
    return {
      index: null,
    }
  },
  computed: {
    visible() {
      if (this.parentAccordion.multipleOpen) {
        return this.parentAccordion.active.includes(this.index)
      }

      return this.index === this.parentAccordion.active
    },
    accordionItemClasses() {
      return ['relative border-b py-2 px-4', { active: this.visible }]
    },
  },
  created() {
    this.index = this.parentAccordion.count++
  },
  methods: {
    hideItem() {
      if (this.parentAccordion.multipleOpen) {
        this.parentAccordion.active.splice(this.parentAccordion.active.indexOf(this.index), 1)
      } else {
        this.parentAccordion.active = null
      }
    },
    showItem() {
      if (this.parentAccordion.multipleOpen) {
        this.parentAccordion.active.push(this.index)
      } else {
        this.parentAccordion.active = this.index
      }
    },
    open() {
      if (this.visible) {
        this.hideItem()
      } else {
        this.showItem()
      }
    },
    start(el) {
      el.style.height = `${el.scrollHeight}px`
    },
    end(el) {
      el.style.height = 'auto'
    },
  },
}
</script>

<style lang="scss" scoped>
.accordion-enter-active,
.accordion-leave-active {
  will-change: height, opacity;
  transition: height 0.3s ease, opacity 0.3s ease;
  overflow: hidden;
}

.accordion-enter,
.accordion-leave-to {
  height: 0 !important;
  opacity: 0;
}

.active {
  .accordion-item-header::after {
    transform: rotate(-180deg) translateY(-50%);
  }
}

.accordion-item-header {
  @apply py-2 cursor-pointer pr-4 relative;

  &::after {
    position: absolute;
    right: 0;
    top: 50%;
    transform: translateY(-50%);
    content: '';
    border-top: 0.325em solid;
    border-right: 0.325em solid transparent;
    border-left: 0.325em solid transparent;
    transition: 0.25s ease;
  }
}
</style>
