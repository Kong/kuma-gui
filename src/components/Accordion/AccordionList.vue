<template>
  <ul class="accordion">
    <slot />
  </ul>
</template>

<script>
export default {
  name: 'AccordionList',
  provide() {
    return { parentAccordion: this.parentAccordion }
  },
  props: {
    initiallyOpen: {
      type: [Number, Array],
      default: null,
    },
    multipleOpen: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    let active
    if (this.initiallyOpen !== null) {
      active = this.initiallyOpen
    } else {
      active = this.multipleOpen ? [] : null
    }

    return {
      parentAccordion: {
        count: 0,
        active,
        multipleOpen: this.multipleOpen,
      },
    }
  },
}
</script>

<style lang="scss" scoped>
.accordion {
  & > li:last-child {
    border-bottom: none;
  }
}
</style>
