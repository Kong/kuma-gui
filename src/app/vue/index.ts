import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
