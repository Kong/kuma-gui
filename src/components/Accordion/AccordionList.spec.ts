import { render, screen } from '@testing-library/vue'
import userEvent from '@testing-library/user-event'
import AccordionList from './AccordionList.vue'
import AccordionItem from './AccordionItem.vue'

const options = {
  slots: {
    default: [
      `
    <AccordionItem>
      <template slot="accordion-header">
        Header 1
      </template>

      <template slot="accordion-content">
        Content 1
      </template>

    </AccordionItem>
    `,
      `
    <AccordionItem>
      <template slot="accordion-header">
        Header 2
      </template>

      <template slot="accordion-content">
        Content 2
      </template>

    </AccordionItem>
    `,
    ],
  },
  stubs: {
    // used to register custom components
    AccordionItem,
  },
}

describe('AccordionList.vue', () => {
  it('renders snapshot at the beginning', () => {
    const { container } = render(AccordionList, options)

    expect(container).toMatchSnapshot()
  })

  it('renders with opened second panel and switch opened panel on click', async () => {
    render(AccordionList, { ...options, props: { initiallyOpen: 1 } })

    expect(screen.queryByText(/Content 1/)).not.toBeInTheDocument()
    expect(screen.getByText(/Content 2/)).toHaveStyle('display: block')

    await userEvent.click(screen.getByText(/Header 1/))

    expect(await screen.findByText(/Content 1/)).toHaveStyle('display: block')
    expect(screen.queryByText(/Content 2/)).not.toBeInTheDocument()
  })

  it('renders initally two opened accordion', async () => {
    render(AccordionList, { ...options, props: { initiallyOpen: [0, 1], multipleOpen: true } })

    expect(screen.getByText(/Content 1/)).toHaveStyle('display: block')
    expect(screen.getByText(/Content 2/)).toHaveStyle('display: block')
  })

  it('renders initally two closed accordions and open it', async () => {
    render(AccordionList, { ...options, props: { multipleOpen: true } })

    expect(screen.queryByText(/Content 1/)).not.toBeInTheDocument()
    expect(screen.queryByText(/Content 2/)).not.toBeInTheDocument()

    await userEvent.click(screen.getByText(/Header 1/))
    await userEvent.click(screen.getByText(/Header 2/))

    expect(await screen.findByText(/Content 1/)).toHaveStyle('display: block')
    expect(await screen.findByText(/Content 2/)).toHaveStyle('display: block')
  })

  it('checks keyboard accessebility', async () => {
    render(AccordionList, { ...options, props: { multipleOpen: true } })

    expect(screen.queryByText(/Content 1/)).not.toBeInTheDocument()

    userEvent.tab()
    expect(screen.queryByText(/Header 1/)).toHaveFocus()

    userEvent.keyboard('[Enter]')

    expect(await screen.findByText(/Content 1/)).toHaveStyle('display: block')
  })
})
