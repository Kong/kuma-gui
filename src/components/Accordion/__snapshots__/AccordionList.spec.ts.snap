// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionList.vue renders snapshot at the beginning 1`] = `
<div>
  <ul
    class="accordion"
  >
    
    
    <li
      class="accordion-item relative border-b"
    >
      <button
        aria-expanded="false"
        class="accordion-item-header"
        type="button"
      >
        
         Header 1 
        
      </button>
      <transition-stub
        appear="false"
        css="true"
        name="accordion"
        persisted="false"
      >
        <!--v-if-->
      </transition-stub>
    </li>
    
    
    <li
      class="accordion-item relative border-b"
    >
      <button
        aria-expanded="false"
        class="accordion-item-header"
        type="button"
      >
        
         Header 2 
        
      </button>
      <transition-stub
        appear="false"
        css="true"
        name="accordion"
        persisted="false"
      >
        <!--v-if-->
      </transition-stub>
    </li>
    
    
  </ul>
</div>
`;
